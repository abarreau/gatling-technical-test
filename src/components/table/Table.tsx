import { FC } from 'react';
import styles from './Table.module.css';
import { EditableField } from './editable-field/EditableField';

/**
 * Component meant to display tabular data, not for design. Number of elements in `datasource.columnHeaders` must match
 * the number of elements in `row.columnValues`.
 * @param datasource The datasource rendered in the table.
 */
export const Table: FC<{ datasource: Readonly<TableDatasource> }> = ({ datasource }) => {
    if(datasource.columnHeaders.length === 0) {
        console.warn('There is no column headers in your datasource.');
        return null;
    }

    return (<table className={styles.table}>
        <thead>
            <tr>
                {
                    datasource.columnHeaders.map(name => <th key={name}>{name}</th>)
                }
            </tr>
        </thead>
        <tbody>
            {
                datasource.rows.length === 0
                    ? <tr>Empty dataset</tr>
                    : datasource.rows.map((row, rowIndex) =>
                        <tr key={rowIndex}>
                            {
                                row.columns.map((column, columnIndex) =>
                                    column.type === 'editable'
                                        ? <EditableField className={styles.editable} key={columnIndex} value={column.value} onValueChange={(val: string) => console.log('Should update !', val)}>
                                            { column.value }
                                            <i className={styles.editIcon}></i>
                                        </EditableField>
                                        : column.type === 'action'
                                            ? <td key={columnIndex}>{
                                                column.buttons.map((button, buttonIndex) =>
                                                    <button key={buttonIndex} onClick={() => button.onClick(button.metadata)}>{ button.label }</button>)
                                            }</td>
                                            : <td key={columnIndex}>{ column.value }</td>
                                )
                            }
                        </tr>
                    )
            }
        </tbody>
    </table>);
};

export type TableDatasource = Readonly<{
    columnHeaders: string[],
    rows: {
        columns: (TableDatasourceActionColumn | TableDatasourceSimpleColumn | TableDatasourceEditableColumn)[]
    }[]
}>;

export type TableDatasourceColumn = Readonly<{
    type: 'editable' | 'simple' | 'action'
    value: string;
}>

export type TableDatasourceEditableColumn = TableDatasourceColumn & Readonly<{
    type: 'editable',
    onEdit?: (value: string) => boolean
}>

export type TableDatasourceSimpleColumn = TableDatasourceColumn & Readonly<{
    type: 'simple'
}>

export type TableDatasourceActionColumn = Pick<TableDatasourceColumn, 'type'> & Readonly<{
    type: 'action',
    buttons: {
        metadata: number;
        label: string;
        onClick: (metadata: number) => void
    }[]
}>