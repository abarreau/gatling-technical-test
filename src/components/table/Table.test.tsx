import renderer from 'react-test-renderer';
import { Table, TableDatasource } from './Table';

describe('Table', () => {
    it('should display component', () => {
        const dummyDatasource: TableDatasource = {
            columnHeaders: [ 'One', 'Two', 'Three' ],
            rows: [
                {
                    columns: [
                        { value: 'Uno', isEditable: false },
                        { value: 'Dos', isEditable: false },
                        { value: 'Tres', isEditable: false }
                    ]
                },
                {
                    columns: [
                        { value: 'Un', isEditable: false },
                        { value: 'Deux', isEditable: false },
                        { value: 'Trois', isEditable: false }
                    ]
                }
            ]
        };

        const component = renderer.create(
            <Table datasource={dummyDatasource}></Table>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should display an empty message if no rows', () => {
        const dummyDatasource: TableDatasource = {
            columnHeaders: [ 'One', 'Two', 'Three' ],
            rows: []
        };

        const component = renderer.create(
            <Table datasource={dummyDatasource}></Table>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should not display anything if there is no column headers', () => {
        console.warn = jest.fn();
        const dummyDatasource: TableDatasource = {
            columnHeaders: [],
            rows: []
        };

        const component = renderer.create(
            <Table datasource={dummyDatasource}></Table>
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        expect(console.warn).toHaveBeenCalledWith('There is no column headers in your datasource.');
    });
});
