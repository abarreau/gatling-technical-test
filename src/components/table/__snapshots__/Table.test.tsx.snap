// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should display an empty message if no rows 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        One
      </th>
      <th>
        Two
      </th>
      <th>
        Three
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      Empty dataset
    </tr>
  </tbody>
</table>
`;

exports[`Table should display component 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        One
      </th>
      <th>
        Two
      </th>
      <th>
        Three
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        Uno
      </td>
      <td>
        Dos
      </td>
      <td>
        Tres
      </td>
    </tr>
    <tr>
      <td>
        Un
      </td>
      <td>
        Deux
      </td>
      <td>
        Trois
      </td>
    </tr>
    <tr>
      <td
        className="editable"
        onClick={[Function]}
      >
        Un
        <i
          className="editIcon"
        />
      </td>
      <td>
        Deux
      </td>
      <td>
        <button
          onClick={[Function]}
        >
          myAction
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table should not display anything if there is no column headers 1`] = `null`;
